cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java

```diff
@@ -173,9 +175,35 @@ private void addMergeOperation(Set<String> stageInputs,
                                              Map<String, Operation> processedOperations) {
     // We need to return InputFields in the same order as fields.
     // Keep them in map so we can iterate later on received fields to return InputFields.
+
     Map<String, InputField> inputFields = new HashMap<>();
     List<String> parents = findParentStages(currentStage);
+
+    List<String> fieldsForJoin = new ArrayList<>();
     for (String field : fields) {
+      if (noMergeRequiredStages.contains(currentStage)) {
+        // Current stage is of type JOIN.
+        // JOIN creates operation with input fields in format <stagename.fieldname>
+        // if the field is directly read from the schema of the one of the input stage.
+        // However it is also possible to have operation with input not tagged with the stage name, for example
+        // in case if rename is done on the join keys. In this case Join operation will read fields of the form
+        // <stagename.fieldname> however rename operations followed by it will simply read the output of the
+        // Join which wont be tagged with stagename.
+
+        //TODO CDAP-13298 Field names and stage names can have '.' in them.
+        // Probably better to support new JOIN operation type which takes stage name and field name.
+        Iterable<String> stageFieldPair = Splitter.on(".").omitEmptyStrings().trimResults().split(field);
+        String stageName = Iterables.get(stageFieldPair, 0);
```

Comment:
I'm not a big fan of the Iterables class. It's usage here can be replaced by getting the iterator, then calling next() on this line, hasNext() instead of Iterables.size(), then next() instead of get(..., 1).

Reply:
fixed.